### PR TITLE
session,server: refactor the execute statement processing

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/16817

Problem Summary:

The root cause of the panic in #16817 is the data race.

And the data race exists because we close the transaction immediately in `ExecuteStmt`.
When using the `RecordSet` returned by `ExecuteStmt`, executor.Next() is called and it maybe still refering the txn, the txn is closed and set to nil however.

https://github.com/pingcap/tidb/blob/0f6b6a2013d9d165589a3a31bb914288854f83b5/session/tidb.go#L211

```
rs :=    ExecuteStmt --->
                runStmt --->
                    finishStmt --->
                        CommitTxn --->
                            txn set to nil !!

rs.Next() -->
    executor.Next --->
        PointGet.Next --->
            txn.Valid() ???
```

### What is changed and how it works?

We should not commit a txn until the executor stop refering it.
The correct order is:

1. ExecuteStmt return a RecordSet `rs`
2. Call `rs.Next()`, drain the chunk data and sent result to the mysql client
3. Call `rs.Close()`
4. Then `finishStmt` should be called and txn maybe commit here.

What's Changed:

- Move `finishStmt` out of `runStmt`.
- Wrap a new `execStmtResult` based on `RecordSet`, override the `Close` method,
and call `finishStmt` in `execStmtResult.Close()`

How it Works:

Txn is not commited until the executor close the RecordSet, so there aren't be any concurrent visits of txn between `finishStmt` and `recordset.Next`, the data race is fixed.

### Related changes

The refactor in https://github.com/pingcap/tidb/pull/16056 avoid the multiple return value `[]RecordSet`. It makes a solid foundantion towards better code for the session processing.

And this commit is a furthur step after that. This commit solve the problem finally:

> The problem is, the recordset is **lazy delivered**, and it references the state of the session which is always mutable.

This commit guarantee the session won't be modify until the chunk data are drained and recordset is closed.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

`runStmtWrap` is a copy of `runStmt`, that's because even a tiny modify would meet numerous CI error, I've tried.

Keep this commit as simple as possible: just move `finishStmt` to `RecordSet.Close`
And after it's merged, I'll clean old `Execute` method (which uses `runStmt`).

### Release note 

- No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
/run-all-tests

PTAL @coocood @jackysp @SunRunAway

**Comment by tiancaiamao** [comment]
PTAL @coocood @jacksp @lysu

**Comment by coocood** [comment]
LGTM

File: session/session.go [hunk 1]
```diff
@@ -1163,7 +1163,7 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
 
 	// Execute the physical plan.
 	logStmt(stmtNode, s.sessionVars)
-	recordSet, err := runStmt(ctx, s, stmt)
+	recordSet, err := runStmtWrap(ctx, s, stmt)
```
Comment (jackysp) [inline session/session.go:1166]: So the binary protocol (execute statement) will still use `runStmt`? Do you need to fix it?

Comment (tiancaiamao) [reply session/session.go:1166]: Yes, it will be handled in the following works.

**Comment by zz-jason** [comment]
@tiancaiamao could you cherry-pick this PR to release 4.0?

**Pull request merged**